### DISCUSSION TITLE
Why entity annotation "decimal" generates assessors using "string" instead of "float"?

### DISCUSSION THREAD

**Discussion by sdespont** [symfony/symfony#43599] [Q&A] [answered]
Hello,

Using decimal precision with Doctrine annotation and specifing `float` for the private variable like this:

```
    /**
     * @ORM\Column(type="decimal", precision=8, scale=2, nullable=false)
     */
    private float $test_float;
```
Generates the following result using the command `php bin/console make:entity --regenerate` :
```
    public function getTestFloat(): ?string
    {
        return $this->test_float;
    }

    public function setTestFloat(string $test_float): self
    {
        $this->test_float = $test_float;

        return $this;
    }
```

I don't understand why the assessors created :

1. are not using `float` for type hinting and return type?
2. the return type is nullable even if option `nullable=false` is set ?
3. Is it safe to change the return type from `string` to `float` like this?

```
    public function getTestFloat(): float
    {
        return $this->test_float;
    }

    public function setTestFloat(float $test_float): self
    {
        $this->test_float = $test_float;

        return $this;
    }
```

Best regards

**Comment by mbabker** [comment, marked as answer]
By design, the DBAL handles decimal and float types differently.  See https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/types.html#decimal-types for more details.

If you have an entity using a decimal field, you need to handle type conversion in your accessors instead of relying on the DBAL and/or ORM to do it for you.

```php
<?php

namespace App\Entity;

use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity]
#[ORM\Table]
class DecimalDemo
{
    #[ORM\Column(nullable: false)]
    #[ORM\Id]
    #[ORM\GeneratedValue]
    private $id;

    #[ORM\Column(type: Types::DECIMAL, precision: 8 scale: 2, nullable: false)]
    private string $decimalValue = '0.00';

    /**
     * @return mixed
     */
    public function getId()
    {
        return $this->id;
    }

    public function getDecimalValue(): float
    {
        return (float) $this->decimalValue;
    }

    public function setDecimalValue(float $decimalValue): void
    {
        $this->decimalValue = sprintf('%.2f', $decimalValue);
    }
}
```

**Reply by sdespont** [reply]
@mbabker Thank you very much for your answer!

After some tests, specifing `float` as type hint to the variable always trigs Doctrine entity update because of the strict comparison failed between `float` and expected `string` for the decimal type. That was a big mistake from my side and your answer help me a lot to understand this.

Because [MySql rounds the number automatically according the decimal configuration](https://dev.mysql.com/doc/refman/8.0/en/precision-math-rounding.html), I don’t plan to use the `sprintf` function as suggested. Also, these kind of function could result with complex situation like [this](https://stackoverflow.com/questions/27107238/why-does-phps-sprintf-not-round-5s-reliably). Finally, because the internal variable is defined as `string`, PHP executes auto casting when assigning the `float` number in the `setTestFloat` method.

At the end, my entity definition is like this:

    /**
     * @ORM\Column(type="decimal", precision=8, scale=2, nullable=false)
     */
    private string $test_float;

    public function getTestFloat(): float
    {
        return $this->test_float;
    }

    public function setTestFloat(float $test_float): self
    {
        $this->test_float = $test_float;

        return $this;
    }

Does this make sense to you?

**Reply by stof** [reply]
if you don't use `number_format` functionto ensure the right precision in the string, you might still trigger extra doctrine updates.

**Reply by sdespont** [reply]
@stof thanks for your comment, so the correct would be 
```
public function setTestFloat(float $test_float): self
{
    $this->test_float = number_format($test_float, 2);

    return $this;
}
```

Is that correct?

**Reply by sdespont** [reply]
@stof Could you please confirm that using `number_format` like below is the right way to handle `float` to `MySql Decimal` format? I really don't find any documentation about this in Symfony documentation 😕 
```
public function setTestFloat(float $test_float): self
{
    $this->test_float = number_format($test_float, 2);
    return $this;
}
```

Thanks

**Reply by stof** [reply]
Yes, it is the right way.

> I really don't find any documentation about this in Symfony documentation

That's expected, as this is not a behavior related to Symfony

**Reply by sdespont** [reply]
Thank you for your reply, I will set the ticket close.

**Reply by sdespont** [reply]
As the behaviour is related to Doctrine, I would be interested about a documentation link because I don't find this recommendation neither in their documentation.
